incompatibility/incompatibility-report/oas3-traverse.go

```diff
@@ -0,0 +1,160 @@
+		return SameType
+	}
+	return DiffType
+}
+
+// arg function should be a single-input single-output function
+// creates generic wrapper around function, indicates error on return
+func makeGenericOperation(function interface{}) (GenericOperation, error) {
+	switch inputType := reflect.TypeOf(function); inputType.Kind() {
+	case reflect.Func:
+		// Continue function
+		validFunction := inputType.NumIn() == 1 && inputType.NumOut() == 1
+		if validFunction {
+			funcInputT := inputType.In(0)
+			funcOutputT := inputType.Out(0)
+			formattedGenericFunction := func(input interface{}) (interface{}, error) {
+				switch t := reflect.TypeOf(input); t {
+				case funcInputT:
+					functionAsValue := reflect.ValueOf(function)
+					returnAsInterface := functionAsValue.Call([]reflect.Value{reflect.ValueOf(input)})[0].Interface()
+					returnType := reflect.TypeOf(returnAsInterface)
+					if returnType != funcOutputT {
+						return nil, TraverseError{"Bad Return Type in Invocation"}
+					}
+					return returnAsInterface, nil
+				default:
+					return nil, TraverseError{"Bad Input Type in Inovocation"}
+				}
+			}
+			return GenericOperation{
+				op:         formattedGenericFunction,
+				inputType:  funcInputT,
+				outputType: funcOutputT,
+			}, nil
+		}
+		goto failGen
+	default:
+		// Invalid input
+		goto failGen
+	}
+failGen:
+	return GenericOperation{}, TraverseError{"input is not a single-input single-output function"}
+}
+
+// function should be a single-input single-output function whose output type is *IncompatibilityReport
+// performs error checking for these requirements and casts into generic type
+func makeGenericIncompatibilityReportFunc(function interface{}) (func(interface{}) (*IncompatibilityReport, error), error) {
+	genGeneric, err := makeGenericOperation(function)
+	badType := genGeneric.outputType != reflect.TypeOf(&IncompatibilityReport{})
+	if err != nil || badType {
+		return nil, err
+	} else {
+		return func(i interface{}) (*IncompatibilityReport, error) {
+			res, err := genGeneric.op(i)
+			return res.(*IncompatibilityReport), err
+		}, nil
+	}
+}
+
+// ComponentType - component type in OpenAPIDoc, i.e. Parameters, Components, ...
+// ParentTraverse - lambda to tranverse from parent to component of type ComponentType
+// OperationOnComponent - lambda which looks at current layer for incompatibilities
+// ChildPaths - paths to check for deeper incompatibilities
+type PathOperation struct {
+	ComponentType        reflect.Type
+	ParentTraverse       GenericOperation
+	OperationOnComponent func(currentComponent interface{}) (*IncompatibilityReport, error)
+	ChildPaths           []*PathOperation
+}
+
+// Merge Two Path Operations under the conditions:
+// 1) Both Operations happen at same component
+//   - Thus compose a new OperationOnComponent
+// 2) p2's parent type is the same as p1's component type
+//   - add p2 to p1 child paths
+func (p1 PathOperation) compose(p2 PathOperation) (PathOperation, error) {
```

Comment:
nit: receivers should be the same name for all methods on a struct. Suggest `p` and `other` for the names.